### PR TITLE
Add derive crate with serialization

### PR DESCRIPTION
**Pull request by limemloh**
## Purpose

Expose smart contract serialization macros outside of smart contracts https://github.com/Concordium/concordium-rust-smart-contracts/issues/146.
Should be review together with https://github.com/Concordium/concordium-rust-smart-contracts/pull/147.

## Changes

- Moved `concordium-contracts-common` into a subfolder.
- Introduced a crate for macros `concordium-contracts-common-derive`.
- Copied serialization macros from `concordium-std-derive` with adjustments for support use in Rust SDK.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Comment by limemloh** [comment]
This PR increases the module size for some examples.
```
 Size  | Diff   | Example
=======+========+================
 21686 |   base | main/auction
 21686 |      0 | contracts-common-derive/auction

 20194 |   base | main/auction-no-std
 20194 |      0 | contracts-common-derive/auction-no-std

 82191 |   base | main/cis2-multi
 82247 |     56 | contracts-common-derive/cis2-multi

 78974 |   base | main/cis2-multi-no-std
 79030 |     56 | contracts-common-derive/cis2-multi-no-std

 79200 |   base | main/cis2-nft
 79256 |     56 | contracts-common-derive/cis2-nft

 75915 |   base | main/cis2-nft-no-std
 75971 |     56 | contracts-common-derive/cis2-nft-no-std

 63678 |   base | main/cis2-wccd
 63734 |     56 | contracts-common-derive/cis2-wccd

 59580 |   base | main/cis2-wccd-no-std
 59636 |     56 | contracts-common-derive/cis2-wccd-no-std

 19958 |   base | main/icecream
 19986 |     28 | contracts-common-derive/icecream

 20420 |   base | main/icecream-no-std
 20448 |     28 | contracts-common-derive/icecream-no-std

 44676 |   base | main/two-step-transfer
 44704 |     28 | contracts-common-derive/two-step-transfer

 40646 |   base | main/two-step-transfer-no-std
 40674 |     28 | contracts-common-derive/two-step-transfer-no-std
```

**Comment by abizjak** [comment]
> This PR increases the module size for some examples. (using rustc 1.62.1 and nightly-2021-06-09-x86_64-unknown-linux-gnu for no_std)
> 
> ```
>  Size  | Diff   | Example
> =======+========+================
>  21686 |   base | main/auction
>  21686 |      0 | contracts-common-derive/auction
> 
>  20194 |   base | main/auction-no-std
>  20194 |      0 | contracts-common-derive/auction-no-std
> 
>  82191 |   base | main/cis2-multi
>  82247 |     56 | contracts-common-derive/cis2-multi
> 
>  78974 |   base | main/cis2-multi-no-std
>  79030 |     56 | contracts-common-derive/cis2-multi-no-std
> 
>  79200 |   base | main/cis2-nft
>  79256 |     56 | contracts-common-derive/cis2-nft
> 
>  75915 |   base | main/cis2-nft-no-std
>  75971 |     56 | contracts-common-derive/cis2-nft-no-std
> 
>  63678 |   base | main/cis2-wccd
>  63734 |     56 | contracts-common-derive/cis2-wccd
> 
>  59580 |   base | main/cis2-wccd-no-std
>  59636 |     56 | contracts-common-derive/cis2-wccd-no-std
> 
>  19958 |   base | main/icecream
>  19986 |     28 | contracts-common-derive/icecream
> 
>  20420 |   base | main/icecream-no-std
>  20448 |     28 | contracts-common-derive/icecream-no-std
> 
>  44676 |   base | main/two-step-transfer
>  44704 |     28 | contracts-common-derive/two-step-transfer
> 
>  40646 |   base | main/two-step-transfer-no-std
>  40674 |     28 | contracts-common-derive/two-step-transfer-no-std
> ```

Can you diff the examples with 56 bytes difference?

Their `.wat` representation. I'm guessing it's paths or some such, and not code size difference.

File: concordium-contracts-common-derive/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+/// A list of valid concordium attributes
+const VALID_CONCORDIUM_ATTRIBUTES: [&str; 1] = ["state_parameter"];
+
+#[cfg(feature = "sdk")]
+fn get_root() -> proc_macro2::TokenStream {
+    quote!(concordium_rust_sdk::types::smart_contracts::concordium_contracts_common)
+}
+
+#[cfg(not(feature = "sdk"))]
+fn get_root() -> proc_macro2::TokenStream { quote!(concordium_std) }
```
Comment (abizjak) [inline concordium-contracts-common-derive/src/lib.rs:28]: Could these not be constants instead?

File: concordium-contracts-common-derive/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+name = "concordium-contracts-common-derive"
+authors = ["Concordium <developers@concordium.com>"]
+version = "0.1.0"
+edition = "2018"
+license = "MPL-2.0"
+description = "Procedural macros to ease writing for smart contracts on the Concordium blockchain."
+readme = "./README.md"
+
+[features]
+sdk = []
```
Comment (abizjak) [inline concordium-contracts-common-derive/Cargo.toml:11]: I do not see any documentation of this feature anywhere. 

Please add it to the readme and explain why it is needed, as well as why this limited scope is chosen.

**Comment by abizjak** [comment]
Also, please add clippy and fmt checks for the new crate with all the feature combinations.

File: concordium-contracts-common-derive/src/lib.rs (hunk 1 above)
Comment (limemloh) [reply concordium-contracts-common-derive/src/lib.rs:28]: I was not able to find a way.

File: concordium-contracts-common-derive/Cargo.toml (hunk 2 above)
Comment (limemloh) [reply concordium-contracts-common-derive/Cargo.toml:11]: I have updated the readme

**Review by abizjak — APPROVED** [review]

**Comment by abizjak** [comment]
We just need to understand the size difference, where it is coming from. I'll look at the examples tonight and we can hopefully get to the bottom of it tomorrwo.

**Pull request merged**